templates/qc_utils/celltype_functions.R

```diff
@@ -0,0 +1,223 @@
+# Create data frame of cell types
+
+
+#' Create `celltype_df` data frame for use in cell type QC reports
+#'
+#' @param processed_sce The processed sce object
+#' @param has_singler Boolean for whether SingleR annotations are present
+#' @param has_cellassign Boolean for whether CellAssign annotations are present
+#'
+#' @return `celltype_df` with column of cell types, as factors, for each annotation method
+create_celltype_df <- function(processed_sce, has_singler, has_cellassign) {
+  celltype_df <- colData(processed_sce) |>
+    as.data.frame() |>
+    # barcodes to a column
+    tibble::rownames_to_column(var = "barcode") |>
+    # keep only cell name, celltyping, and clusters
+    dplyr::select(
+      barcode,
+      clusters,
+      contains("singler"),
+      contains("cellassign"),
+      contains("submitter")
+    )
+
+  if (has_singler) {
+    celltype_df <- celltype_df |>
+      prepare_annotation_values(singler_celltype_annotation)
+  }
+  if (has_cellassign) {
+    celltype_df <- celltype_df |>
+      prepare_annotation_values(cellassign_celltype_annotation)
+  }
+
+  return (celltype_df)
+}
+
+# Functions for working with cell type annotations in QC reports
+
+#' Prepare and reformat cell type annotation values for use in QC reports
+#'  Unknown cell types are updated with the label "Unknown cell type", and
+#'  cell types are ordered in order of descending frequency, but with
+#'  "Unknown cell type" as the last level
+#'
+#' @param df The data frame containing cell type annotations, one row per cell
+#' @param annotation_column The column (written plainly, not a string) containing annotations to reformat
+#'
+#' @return Updated data frame with the `annotation_column` reformatted
+prepare_annotation_values <- function(df, annotation_column) {
+  df |>
+    dplyr::mutate(
+      {{ annotation_column }} := dplyr::case_when(
+        # singler condition
+        is.na({{ annotation_column }}) ~ "Unknown cell type",
+        # cellassign conditon
+        {{ annotation_column }} == "other" ~ "Unknown cell type",
+        # otherwise, keep it
+        TRUE ~ {{ annotation_column }}
```

Comment:
updated case_when syntax
```suggestion
        .default = {{ annotation_column }}
```